File: skills/clawhub/embodied-bidding-tracker/src/utils/llm.js
/**
 * LLM 调用封装
 *
 * 发布版只保留 OpenAI-compatible HTTP provider，避免在 ClawHub 发布包中包含
 * 本地 CLI 执行能力。真实 API key 只写入本地的 config/settings.json，该文件
 * 已被 .gitignore / .clawhubignore 排除。
 */

import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import path from 'path';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const projectRoot = path.resolve(__dirname, '../..');

function loadConfig() {
  const raw = readFileSync(path.join(projectRoot, 'config', 'settings.json'), 'utf-8');
  const cfg = JSON.parse(raw).llm;
  if (cfg.provider !== 'openai-compatible') {
    throw new Error('发布版仅支持 llm.provider = "openai-compatible"');
  }
  return cfg.openaiCompatible ?? {};
}

async function callOpenAICompatible(systemPrompt, userContent, cfg) {
  const { baseURL, apiKey, model, maxTokens } = cfg;

  if (!baseURL || !model) {
    throw new Error('openaiCompatible.baseURL / model 未配置，请检查 config/settings.json');
  }
  if (!apiKey || apiKey.includes('REPLACE_WITH')) {
    throw new Error('openaiCompatible.apiKey 未配置，请在本地 config/settings.json 中填写 API key');
  }

  const res = await fetch(`${baseURL}/chat/completions`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      model,
      max_tokens: maxTokens ?? 2000,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userContent },
      ],
    }),
  });

  if (!res.ok) {
    const errText = await res.text();
    throw new Error(`OpenAI 兼容接口返回 ${res.status}: ${errText.slice(0, 300)}`);
  }

  const data = await res.json();
  return data.choices[0].message.content.trim();
}

/**
 * 调用 LLM
 * @param {string} systemPrompt  系统提示词
 * @param {string} userContent   用户内容（原始页面文本等）
 * @returns {Promise<string>}    模型输出文本
 */
export async function callLLM(systemPrompt, userContent) {
  return callOpenAICompatible(systemPrompt, userContent, loadConfig());
}

File: skills/clawhub/feeds-digest/scripts/lib/llm.py
"""LLM summarizer for feed digests."""
from __future__ import annotations

import os
from pathlib import Path

import requests


class LLMSummaryError(Exception):
    """Raised when LLM summarization fails."""


def summarize(
    digest_markdown: str,
    provider: str,
    model: str,
    prompt_template: str,
    max_tokens: int = 800,
    temperature: float = 0.2,
) -> str:
    """Generate an LLM summary from the digest."""
    if provider == "perplexity":
        return _call_perplexity(
            digest_markdown, model, prompt_template, max_tokens, temperature
        )
    elif provider == "openai":
        return _call_openai(
            digest_markdown, model, prompt_template, max_tokens, temperature
        )
    elif provider == "ollama":
        return _call_ollama(
            digest_markdown, model, prompt_template, max_tokens, temperature
        )
    else:
        raise LLMSummaryError(f"Unbekannter LLM-Provider: {provider}")


def _call_perplexity(
    digest: str, model: str, prompt_template: str, max_tokens: int, temperature: float
) -> str:
    api_key = os.environ.get("PERPLEXITY_API_KEY")
    if not api_key:
        raise LLMSummaryError("PERPLEXITY_API_KEY nicht gesetzt")

    url = "https://api.perplexity.ai/chat/completions"
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
    }
    payload = {
        "model": model,
        "messages": [
            {"role": "system", "content": prompt_template},
            {"role": "user", "content": digest[:15000]},  # truncate
        ],
        "max_tokens": max_tokens,
        "temperature": temperature,
    }

    try:
        r = requests.post(url, json=payload, headers=headers, timeout=60)
        r.raise_for_status()
        data = r.json()
        return data["choices"][0]["message"]["content"]
    except requests.RequestException as e:
        raise LLMSummaryError(f"Perplexity API-Fehler: {e}") from e
    except (KeyError, IndexError) as e:
        raise LLMSummaryError(f"Unerwartete Perplexity-Response: {e}") from e


def _call_openai(
    digest: str, model: str, prompt_template: str, max_tokens: int, temperature: float
) -> str:
    api_key = os.environ.get("OPENAI_API_KEY")
    if not api_key:
        raise LLMSummaryError("OPENAI_API_KEY nicht gesetzt")

    url = "https://api.openai.com/v1/chat/completions"
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
    }
    payload = {
        "model": model,
        "messages": [
            {"role": "system", "content": prompt_template},
            {"role": "user", "content": digest[:15000]},
        ],
        "max_tokens": max_tokens,
        "temperature": temperature,
    }

    try:
        r = requests.post(url, json=payload, headers=headers, timeout=60)
        r.raise_for_status()
        data = r.json()
        return data["choices"][0]["message"]["content"]
    except requests.RequestException as e:
        raise LLMSummaryError(f"OpenAI API-Fehler: {e}") from e
    except (KeyError, IndexError) as e:
        raise LLMSummaryError(f"Unerwartete OpenAI-Response: {e}") from e


def _call_ollama(
    digest: str, model: str, prompt_template: str, max_tokens: int, temperature: float
) -> str:
    ollama_host = os.environ.get("OLLAMA_HOST", "http://localhost:11434")
    url = f"{ollama_host}/api/generate"
    payload = {
        "model": model,
        "prompt": f"{prompt_template}\n\n---\n\n{digest[:15000]}",
        "stream": False,
        "options": {"temperature": temperature, "num_predict": max_tokens},
    }

    try:
        r = requests.post(url, json=payload, timeout=120)
        r.raise_for_status()
        data = r.json()
        return data.get("response", "")
    except requests.RequestException as e:
        raise LLMSummaryError(f"Ollama API-Fehler: {e}") from e


def load_prompt_template(prompt_file: str | Path) -> str:
    """Load prompt template from file."""
    path = Path(prompt_file)
    if not path.exists():
        # Try relative to skill root
        skill_root = Path(__file__).resolve().parent.parent.parent
        path = skill_root / prompt_file

    if not path.exists():
        raise LLMSummaryError(f"Prompt-Template nicht gefunden: {prompt_file}")

    with open(path, encoding="utf-8") as f:
        return f.read()
